3rdparty/obiwarp/dynprog.cpp

```diff
@@ -1062,20 +1062,20 @@ float entropy(MatF &mat, int rowNum, int numBins, float minVal, float scaleFacto
 
 //Subtract a value
 void _subtract(MatF &mat, int rowNum, float val, MatF &minused) {
-    std::vector<float> matptr = mat.pointer(rowNum);
-    std::vector<float> minusedptr = minused.pointer(rowNum);
+    float* matPtr = mat.rowData(rowNum);
```

Comment:
We can talk in person if you think that would be easier

Reply:
The STL `vector` container provides a lot of nice methods but is in fact simply an abstraction over a contiguous block of elements (read "array"). Its [`data`](https://en.cppreference.com/w/cpp/container/vector/data) method allows the programmer to access the underlying array by returning a pointer to the first element. This is what `rowData` method for the matrix classes returns (and that is why their return types are pointers of `float`, `double` or `int` depending on the type of matrix).

The `rowIters` method does not return a pointer but iterators to the first and last element of a row in a matrix. Since a matrix is internally just a flat vector structure (with rows arranged adjacent to each other), you can directly access an element by using pointers or iterators, which also give you the guarantee that if you move ahead by 1 you will reach the next element in the row, and therefore, enable looping over the entire row (just make sure you stop at the end of the row by either checking for the size traversed or by checking if your iterator points to the last element at each step).

Does that make sense? I might have complicated this a bit so let me know if you want to talk about this in person.

Reply:
Makes perfect sense. Thanks!